src/department/department.service.ts

```diff
@@ -2,48 +2,77 @@ import { CreateDepartmentDto } from "./dto/createDepartment.dto";
 
 @Injectable()
 export class DepartmentSevice {
   constructor(
     @InjectRepository(DepartmentEntity)
     private readonly departmentRepository: Repository<DepartmentEntity>
   ) {}
-  async getAll(): Promise<DepartmentEntity[]> {
-    return await this.departmentRepository.find();
+  async getAll(
+    query: DepartmentsOptionInterface
+  ): Promise<DepartmentsResponseInterface> {
+    const queryBuilder =
+      getRepository(DepartmentEntity).createQueryBuilder("department");
+    const departmentsCount = await queryBuilder.getCount();
+    queryBuilder.orderBy("department.name", "ASC");
+    if (query.name) {
+      queryBuilder.andWhere("department.name = :name", { name: query.name });
+    }
+    if (query.limit) {
+      queryBuilder.limit(query.limit);
+    }
+    if (query.offset) {
+      queryBuilder.offset(query.offset);
+    }
+    const departments = await queryBuilder.getMany();
+    return { count: departmentsCount, data: departments };
   }
   async getById(id: number): Promise<DepartmentEntity> {
-    return await this.departmentRepository.findOne({ id });
+    const department = await this.departmentRepository.findOne({ id });
+    if (!department) {
+      throw new HttpException(
+        "The department with this ID was not found.",
+        HttpStatus.NOT_FOUND
+      );
+    }
+    return department;
   }
   async createDepartment(
     createDepartmentDto: CreateDepartmentDto
   ): Promise<DepartmentEntity> {
-    const newDepartment: DepartmentEntity = new DepartmentEntity();
+    const newDepartment = new DepartmentEntity();
     Object.assign(newDepartment, createDepartmentDto);
     newDepartment.employees = [];
+    const сheckName = await this.departmentRepository.findOne({
+      name: newDepartment.name,
+    });
+    if (сheckName) {
+      throw new HttpException(
+        "The department with this Name already exists.",
+        HttpStatus.BAD_REQUEST
+      );
+    }
     return await this.departmentRepository.save(newDepartment);
   }
   async updeteById(
     id: number,
     updateDepartmentDto: UpdateDepartmentDto
   ): Promise<DepartmentEntity> {
-    const department: DepartmentEntity = await this.getById(id);
-    if (!department) {
-      throw new HttpException(
-        "Department does not exist",
-        HttpStatus.NOT_FOUND
-      );
-    }
+    const department = await this.getById(id);
     Object.assign(department, updateDepartmentDto);
     return await this.departmentRepository.save(department);
   }
   async deleteById(id: number): Promise<DeleteResult> {
-    const department: DepartmentEntity = await this.getById(id);
-    if (!department) {
+    const department = await this.getById(id);
+
+    if (department.employees.length !== 0) {
       throw new HttpException(
-        "Department does not exist",
+        "Unable to delete a department. The department contains employees.",
         HttpStatus.NOT_FOUND
```

Comment:
this should be put onto db constraints. ON DELETE RESTRICT instead
https://orkhan.gitbook.io/typeorm/docs/relations#relation-options